### PR TITLE
Plugin system for Ottoman

### PR DESCRIPTION
**Pull request by moxious**
This commit introduces a plugin system for Ottoman.  Plugins are tiny functions that take model instances and options as arguments.  They use the eventing system (pre/post save, load, etc) to perform useful functions.

A plugin system is an important piece to have because it lets users extend ottoman functionality with stuff that doesn't really belong in Ottoman proper.

This commit comes with unit tests covering 100% of the new code introduced.

Finally -- the setup of this plugin system is based on Mongoose, which has a decent ecosystem for plugins, and a very simple implementation.  Below are some links showing correspondence between what I did in this commit and what's in mongoose's source code.  The code is pretty self-explanatory, but the parallels between the two are clear (and intentional).

[Mongoose - how to register global plugins](https://github.com/Automattic/mongoose/blob/master/lib/index.js#L450)

[Mongoose plugins on model instances](https://github.com/Automattic/mongoose/blob/master/lib/schema.js#L962)

[Mongoose application of global plugins to created models](https://github.com/Automattic/mongoose/blob/master/lib/index.js#L431)

### CODE REVIEW

**Comment by moxious** [comment]
Note this PR includes commits which are requested as part of this other PR:  https://github.com/couchbaselabs/node-ottoman/pull/87   

Recommend reviewing that PR #87 first.    There won't be any merge conflicts between the two in any case though.

File: lib/ottoman.js [hunk 1]
```diff
@@ -736,10 +758,23 @@ Ottoman.prototype._buildAndRegisterModel = function(name, schemaDef, options) {
...
   }
 
+  this._applyPlugins(model);
+
   return model;
 };
 
 /**
+ * Applies global plugins to models as they are created.
+ * @param a new model that has been created.
```
Comment (brett19) [inline lib/ottoman.js]: The first part of an @param should be the name of the parameter.

**Comment by brett19** [comment]
Hey @moxious ,
This looks good, though the commit contains a few whitespace fix-ups which should go in their own commit (a big commit of just fixing up the code style).
Cheers, Brett

**Comment by moxious** [comment]
I'll see if I can split up the whitespace fixes.  Will act on that other minor jsdoc comment too

**Comment by moxious** [comment]
JSdoc thingy fixed.  Whitespace is pretty difficult for me to split out, in part because I'm running vscode with various plugins, the editor doesn't really want to let me save without some of those minor changes.

If desired, I can do a bulk cleanup and submit a "whitespace/format" only PR to prettify with the editorconfig setup as I've tried to stay limited to line ranges so far and I may have just slipped here.  It'd make it easier for me to hack on ottoman but isn't necessarily important for anyone else.

**Comment by brett19** [comment]
Hey @moxious,
If you do a big cleanup on its own branch off master, I can pull that in and then rebase this on top of it, which will remove the whitespace changes from this specific commit.
Cheers, Brett

**Comment by brett19** [comment]
Merged.

**Pull request closed without merging**